File: 25-tables/csharp/Tables/Tables.cs
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;
using System;
using System.Collections.Generic;

public class Tables
{
    IWebDriver Driver;

    [SetUp]
    public void SetUp()
    {
        Driver = new FirefoxDriver();
    }

    [TearDown]
    public void TearDown()
    {
        Driver.Quit();
    }

    [Test]
    public void TableWithNoHelpfulMarkup1()
    {
        Driver.Navigate().GoToUrl("http://the-internet.herokuapp.com/tables");

        // sort "Due" column in ascending order
        Driver.FindElement(By.CssSelector("#table1 thead tr th:nth-of-type(4)")).Click();

        // get values from "Due" column and clean up the data
        IReadOnlyCollection<IWebElement> Dues = Driver.FindElements(By.CssSelector("#table1 tbody tr td:nth-of-type(4)"));
        List<double> FormattedDues = new List<double>();
        foreach(IWebElement Due in Dues)
        {
            FormattedDues.Add(double.Parse(Due.Text.Replace("$", "")));
        }

        // check that the "Due" column is in ascending order
        Assert.That(FormattedDues, Is.Ordered);
    }

    [Test]
    public void TableWithNoHelpfulMarkup2()
    {
        Driver.Navigate().GoToUrl("http://the-internet.herokuapp.com/tables");

        // sort "Due" column in descending order
        Driver.FindElement(By.CssSelector("#table1 thead tr th:nth-of-type(4)")).Click();
        Driver.FindElement(By.CssSelector("#table1 thead tr th:nth-of-type(4)")).Click();

        // get values from "Due" column and clean up the data
        IReadOnlyCollection<IWebElement> Dues = Driver.FindElements(By.CssSelector("#table1 tbody tr td:nth-of-type(4)"));
        List<double> FormattedDues = new List<double>();
        foreach(IWebElement Due in Dues)
        {
            FormattedDues.Add(double.Parse(Due.Text.Replace("$", "")));
        }

        // check that the "Due" column is in ascending order
        Assert.That(FormattedDues, Is.Ordered.Descending);
    }

    [Test]
    public void TableWithNoHelpfulMarkup3()
    {
        Driver.Navigate().GoToUrl("http://the-internet.herokuapp.com/tables");

        // sort "Email" column in descending order
        Driver.FindElement(By.CssSelector("#table1 thead tr th:nth-of-type(3)")).Click();

        // get values from "Email" column
        IReadOnlyCollection<IWebElement> Emails = Driver.FindElements(By.CssSelector("#table1 tbody tr td:nth-of-type(3)"));
        List<string> FormattedEmails = new List<string>();
        foreach(IWebElement Email in Emails)
        {
            FormattedEmails.Add(Email.Text);
        }

        // check that the "Email" column is in ascending order
        Assert.That(FormattedEmails, Is.Ordered);
    }

    [Test]
    public void TableWithHelpfulMarkup()
    {
        Driver.Navigate().GoToUrl("http://the-internet.herokuapp.com/tables");
        Driver.FindElement(By.CssSelector("#table2 thead .dues")).Click();
        IReadOnlyCollection<IWebElement> Dues = Driver.FindElements(By.CssSelector("#table2 tbody .dues"));
        List<double> FormattedDues = new List<double>();
        foreach(IWebElement Due in Dues)
        {
            FormattedDues.Add(double.Parse(Due.Text.Replace("$", "")));
        }
        Assert.That(FormattedDues, Is.Ordered);
    }
}

File: 25-tables/python/tables.py
# -*- coding: utf-8 -*-
"""
Implementation of http://elementalselenium.com/tips/25-tables
"""

import unittest
from selenium import webdriver


class Tables(unittest.TestCase):

    def setUp(self):
        self.driver = webdriver.Firefox()

    def tearDown(self):
        self.driver.quit()

    def test_sort_number_column_in_ascending_order_with_limited_locators(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com/tables')
        driver.find_element_by_css_selector('#table1 thead tr th:nth-of-type(4)').click()
        due_column = driver.find_elements_by_css_selector('#table1 tbody tr td:nth-of-type(4)')
        dues = [float(due.text.replace('$','')) for due in due_column]
        assert dues == sorted(dues)

    def test_sort_number_column_in_descending_order_with_limited_locators(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com/tables')
        driver.find_element_by_css_selector('#table1 thead tr th:nth-of-type(4)').click()
        driver.find_element_by_css_selector('#table1 thead tr th:nth-of-type(4)').click()
        due_column = driver.find_elements_by_css_selector('#table1 tbody tr td:nth-of-type(4)')
        dues = [float(due.text.replace('$','')) for due in due_column]
        assert dues == sorted(dues, reverse=True)

    def test_sort_text_column_in_ascending_order_with_limited_locators(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com/tables')
        driver.find_element_by_css_selector('#table1 thead tr th:nth-of-type(3)').click()
        email_column = driver.find_elements_by_css_selector('#table1 tbody tr td:nth-of-type(3)')
        emails = [email.text for email in email_column]
        assert emails == sorted(emails)

    def test_sort_number_column_in_ascending_order_with_helpful_locators(self):
        driver = self.driver
        driver.get('http://the-internet.herokuapp.com/tables')
        driver.find_element_by_css_selector('#table2 thead .dues').click()
        due_column = driver.find_elements_by_css_selector('#table2 tbody .dues')
        dues = [float(due.text.replace('$','')) for due in due_column]
        assert dues == sorted(dues)

if __name__ == "__main__":
    unittest.main()
